src/Mvc/Mvc.Core/src/Formatters/SystemTextJsonOutputFormatter.cs

```diff
@@ -68,11 +74,22 @@ public sealed override async Task WriteResponseBodyAsync(OutputFormatterWriteCon
 
         var httpContext = context.HttpContext;
 
-        // context.ObjectType reflects the declared model type when specified.
-        // For polymorphic scenarios where the user declares a return type, but returns a derived type,
-        // we want to serialize all the properties on the derived type. This keeps parity with
-        // the behavior you get when the user does not declare the return type and with Json.Net at least at the top level.
-        var objectType = context.Object?.GetType() ?? context.ObjectType ?? typeof(object);
+        // Maybe we could use the jsontypeinfo overload but we need the untyped,
+        // waiting for https://github.com/dotnet/runtime/issues/77051
+
+        var declaredType = context.ObjectType;
+        var runtimeType = context.Object?.GetType();
+        var objectType = runtimeType ?? declaredType ?? typeof(object);
+
+        if (declaredType is not null &&
+            runtimeType != declaredType &&
+            SerializerOptions.TypeInfoResolver != null &&
```

Comment:
`TypeInfoResolver` should be non-null at this point.

Reply:
The serializer could not be called yet, so, null still valid, no?

Reply:
`null` resolvers only get accepted by the serialization methods that are marked `RequiredUnreferencedCode`. Given that you plan on replacing these with the linker-safe overloads, your code will sooner or later need to enforce that invariant.

Reply:
> Given that you plan on replacing these with the linker-safe overloads, your code will sooner or later need to enforce that invariant.

That is true for native AOT (Trim safe)-only. When the app is not configured to trim it could fall back to the current call to the overloaded marked with `RequiredUnreferencedCode`.

Reply:
I think I finally got your point 😂. If the `TypeInfoResolver` is `null` at this point is an indication of:

1. It was explicitly set to `null` by the users, since we are setting it to `Default...` in a non-trimmed(aot) apps, or;
2. It is a trimmed/aot app and the source gen context was not configured correctly.

If I understand correctly, in both case a call to `MakeReadOnly` will throw because it is misconfigured, that really make sense, and that is why you are asking for make the call very early. Is that correct? 

If so, I believe you are correct, and we should only make calls to linker-safe overloads both here and RDF (doesn't matter if in AOT/Trimmed app or not) and add the `MakeReadOnly` as early as possible.